keps/sig-architecture/4330-compatibility-versions/README.md

```diff
@@ -0,0 +1,956 @@
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+We intend to introduce a "compatibility version" option to Kubernetes control
+plane components to make upgrades safer by increasing the granularity of steps
+available to cluster administrators. The "compatibility version" is distinct from
+the binary version of a Kubernetes component, and can be used to emulate the
+behavior (APIs, features, ...) of a prior Kubernetes release version.
+
+## Motivation
+
+The notion of more granular steps in Kubernetes upgrades is attractive because
+it is more rigorous about how we step through a Kubernetes control-plane
+upgrade, introducing potentially corrupting data (i.e. data only present in N+1
+and not in N) only in later stages of the upgrade process.
+
+For example, upgrading from Kubernetes 1.30 to 1.31 while keeping the compatibility
+version at 1.30 would enable a cluster administrator to validate that the new
+Kubernetes binary version is working as desired before exposing any feature changes
+introduced in 1.31 to cluster users, and without writing and data to storage at
+newer API versions.
+
+This extra step increases the granularity of our upgrade sequence so that 
+(1) failures are more easily diagnosed (since we have more granular steps, we
+have slightly more data sheerly from which point of the upgrade lifecycle we
+achieved failure) and (2) failures are more easily auto-reverted by upgrade
+orchestration as we are taking smaller and more incremental steps forward,
+which means there is less to “undo” on a failure condition.
+
+In beta, we intend to introduce support for greater compatibility version skew
+(specifically, N-3) so that it would be possible to skip-level upgrade a 
```

Comment:
imo the skip-level upgrade may be a wrong jargon in this sequence. May be the right word is step level upgade. 

Say we put on hats of a cluster admin responsible for setting up cluster lifecycle platform. The users/admin of the cluster is responsible for upgrading the clusters. This sequence adds safety guarantee to the operation, but does not really skip versions. The number of operations do increase for good reasons.

Reply:
Good catch. Yes, "skip-level" is adding confusion here. I'll rephrase.